### ISSUE TITLE
"The actor's namespace"

### ISSUE THREAD

**Issue by evanp** [w3c-social/activitypump#443]
@trwnh notes in #342 that we specify that the server should create an ID "in the actor's namespace".

"The actor's namespace" isn't a meaningful term in AP. It implies that there are ways to tell which actor created an object by inspecting the ID. Not great!

I believe we need an erratum for this issue.

**Comment by trwnh** [comment]
revisiting this, it feels like the kind of thing that could actually enable actors to be more portable; by making objects relative to "the actor's namespace", you have the option to treat that "namespace" as a gateway, proxy, or redirector to some final location.

this does seem to require that we give up on "reparenting" objects to different actors, but we could maybe use `Move.object == Move.actor` instead of `Update.object.attributedTo` for that. http 3xx can also work as long as the old location's server is up.

at least in a domain-based same-origin-based network, the alternative to "the actor's namespace" is "the server's namespace", i.e. its FQDN (likewise, we assume that a server cannot live in a subpath of a FQDN)

**Comment by trwnh** [comment]
> at least in a domain-based same-origin-based network, the alternative to "the actor's namespace" is "the server's namespace", i.e. its FQDN (likewise, we assume that a server cannot live in a subpath of a FQDN)

3.1 Object Identifiers contains this text currently:

> Publicly dereferencable URIs, such as HTTPS URIs, with their authority belonging to that of their originating server. (Publicly facing content SHOULD use HTTPS URIs).

which maybe should be reconsidered alongside this issue? at least the "authority belonging to that of their originating server" bit. this bit gives the server implicit authority over all objects, which might not be desirable architecturally speaking. an object might be created and given an identifier within "the actor's namespace" to indicate that the object is owned by the actor and not by the server. (this requires the server to be aware of "the actor's namespace", but that's not hard.)

basically if we have a server `https://activitypub.example` and an actor on that server, we can let the server know that "the actor's namespace" is actually `https://trwnh.com/activitypub/` and then it will assign identifiers within that namespace. later when the actor migrates to `https://cooler-activitypub.example` the ids don't change. the only thing we'd have to do is set up an http url rewriter and redirector from `https://trwnh.com/activitypub/` to whatever the current activitypub server is.

**Comment by evanp** [comment]
I added this [erratum](https://www.w3.org/wiki/ActivityPub_errata/Proposed):

> Section 3.1, "Object identifiers", should read in part: However, for client to server communication, a server receiving an object posted to the outbox with no specified id SHOULD allocate an object ID and attach it to the posted object.

It leaves out the mention of "the actor's namespace".

**Comment by evanp** [comment]
We did a CFC for this issue:

https://lists.w3.org/Archives/Public/public-swicg/2025Nov/0015.html

It should close by 12 Dec 2025.

**Comment by snarfed** [comment]
LGTM.

(I'm with @trwnh that adding the concept of actor namespace paths _could_ be interesting and useful!...but it'd be a big addition, and would move URL design from comfortably out of scope to largely in scope, and would open a bunch of other new questions that make it way too big for an erratum or anything else short of a full WG.)

**Comment by bumblefudge** [comment]
> revisiting this, it feels like the kind of thing that could actually enable actors to be more portable; by making objects relative to "the actor's namespace", you have the option to treat that "namespace" as a gateway, proxy, or redirector to some final location.
> 
> this does seem to require that we give up on "reparenting" objects to different actors, but we could maybe use `Move.object == Move.actor` instead of `Update.object.attributedTo` for that. http 3xx can also work as long as the old location's server is up.
> 
> at least in a domain-based same-origin-based network, the alternative to "the actor's namespace" is "the server's namespace", i.e. its FQDN (likewise, we assume that a server cannot live in a subpath of a FQDN)

Note that the DID method spec DOES define something very similar conceptually to the "actor's namespace": the "DID's namespace" (or DID controller's namespace, to better make the analogy explicit).  This namespace would refer to all the resources dereferenceable from a [special kind of URL called a "DID URL" colloquially](https://www.w3.org/TR/did-1.0/#did-url-syntax), which "hangs off of" the DID as authority. These were specified in 2017~8, so would have been known to the AP WG at the time of drafting. (Ironically, around the time DID concepts were being removed from the AP spec for lack of normref, the DID WG was actually [adding some reserve words](https://github.com/w3c/did/commit/957ef133d1ff7cf703bc6e99b1a61b3191478628) for the express purpose of constraining each DID's "namespace"). A high-level/generic syntax is defined in the DID spec for DID-consuming systems to be familiar with, but they are optional for each method to specify, and relatively few DID methods actually implement this URL syntax. The most common usage is for DIDComm purposes, which is mostly off-topic here.  

Note, however, that the ATProto permalink syntax, designed to enable swappable service-providers for data hosting and long-lived URLs that survive their replacement, looks like this:
`at://<did:{plc|web}>/<lexicon>/<rkey>`
If you squint a little, or know the history of the ATProto design process, you might recognize the above as a more intuitive and ergonomic transformation of the older
`<did:method:identifier>/<service>?key=<rkey>`
syntax used by the [older DIF PDS specification](https://github.com/decentralized-identity/decentralized-web-node/blob/2434c10221360a7fb12a5808db9f7b956c24f69e/spec/spec.md#did-relative-urls) from around the time the BSky time was doing their prior art survey. (Note that this DID URL syntax also got abstracted over in a very different way by the [later versions](https://identity.foundation/decentralized-web-node/spec/#messages) of the DIF PDS, renamed "decentralized web node" by the company then driving it, Block/TBD.)

If the above was a little esoteric or needlessly detailed, I could summarize it as follows: insofar as DID URLs are a thing, a "DID's namespace" is all records that can be resolved via that DID's resolution mechanisms from a DID-as-authority URL (in general, as in ATProto, resolving the DID and transforming or redirecting those resource links to the "current PDS host" results in traditional, non-portable URLs). In some early drafts of AP, actors were very DID-like, so this is probably what was "meant" by the phrase, "actor's namespace."

Whether or not it's worth keeping this vestigial trace of a more nuanced C2S in v1.1, is a question for the Working Group. But I would definitely advocate for elaborating a coherent and normatively-anchored vision of C2S and actor namespaces in future.

**Comment by evanp** [comment]
There was feedback from @thisismissem on this issue that the change would not allow transient activities to be created using the API:

https://lists.w3.org/Archives/Public/public-swicg/2025Nov/0018.html

However, since this is not a new change, we added a new ticket to track this issue:

https://github.com/w3c/activitypub/issues/525

Per @bumblefudge , there are URI formats that could incorporate the idea of a "an actor's namespace", but we don't have a way to define it in the AP spec specifically for HTTPS URLs.

Based on this feedback, I have added the erratum and applied it to the editor's draft. We can continue work on closely related issues in #525 and in the future version of AP.

**Comment by trwnh** [comment]
## for next version

here's an example of how "the actor's namespace" could ease portability

### before: i am hosted on mastodon.social

i configure my web server (nginx) like so:

```nginx
location ~ ^/statuses/(.*) {
  return 307 https://mastodon.social/users/trwnh/statuses/$1 ;
} 
```

requesting my authoritative origin redirects to a document whose id is what was authoritatively requested (authenticated by TLS)

```http
GET /statuses/123 HTTP/1.1
Host: trwnh.com

HTTP/1.1 307 Temporary Redirect
Location: https://mastodon.social/users/trwnh/statuses/123

GET /users/trwnh/statuses/123 HTTP/1.1
Host: mastodon.social

HTTP/1.1 200 OK
Content-Type: application/activity+json

{
  "id": "https://trwnh.com/statuses/123"
}
```

### after: i am hosted on cooler-mastodon.social

i export my data from mastodon.social and upload it to cooler-mastodon.social, then i make one change in my web server's configuration:

```nginx
location ~ ^/statuses/(.*) {
  return 307 https://cooler-mastodon.social/users/trwnh/statuses/$1 ;
} 
```

from the outside, there is no change -- requesting a resource from trwnh.com returns a document whose id matches the requested id on trwnh.com

```http
GET /statuses/123 HTTP/1.1
Host: trwnh.com

HTTP/1.1 307 Temporary Redirect
Location: https://cooler-mastodon.social/users/trwnh/statuses/123

GET /users/trwnh/statuses/123 HTTP/1.1
Host: cooler-mastodon.social

HTTP/1.1 200 OK
Content-Type: application/activity+json

{
  "id": "https://trwnh.com/statuses/123"
}
```

### analysis

in the above example before-and-after, "the actor's namespace" is `https://trwnh.com/`. the resource's relative reference is `statuses/123`. as long as mastodon.social and cooler-mastodon.social recognize that namespace and bind it to my actor, they can correctly serve the requested information.

if mastodon.social assigned identifiers in its *own* namespace, "the server's namespace" instead of "the actor's namespace", then portability would be much harder because the identifier would have to be rewritten and/or redirected between mastodon.social and cooler-mastodon.social, without my actor being involved at all. say for example an id was `https://mastodon.social/ap/objects/123` with no relation to `https://mastodon.social/ap/users/14715`, and no equivalences drawn to `https://cooler-mastodon.social/ap/objects/1` and `https://cooler-mastodon.social/ap/users/1`. it wouldn't be easy to do with regex! you would need at least a map from old ids to new ids, and everyone involved would have to rewrite their stored ids when encountering a permanent redirect or receiving a Move or some other equivalent mechanism.
